content/blog/datafusion-custom-parquet-index.md

```diff
@@ -0,0 +1,232 @@
+) -> Result<()> {
+    // Get current byte offset (end of data pages)
+    let offset = writer.bytes_written();
+
+    // Build newline‑separated UTF‑8 payload
+    let serialized = index.inner.iter().cloned().collect::<Vec<_>>().join("\n");
+    let bytes = serialized.into_bytes();
+
+    // Write magic + length
+    writer.write_all(INDEX_MAGIC)?;
+    writer.write_all(&(bytes.len() as u64).to_le_bytes())?;
+
+    // Write index payload
+    writer.write_all(&bytes)?;
+
+    // Append offset metadata to footer
+    writer.append_key_value_metadata(
+        KeyValue::new(
+            "distinct_index_offset".to_string(),
+            offset.to_string(),
+        ),
+    );
+    Ok(())
+}
+```
+
+### Reading the Index
+
+On the read path, we:
+
+1. Open the Parquet footer and extract `distinct_index_offset`.
+2. Seek to that offset in the file.
+3. Read and validate `IDX1` magic.
+4. Read the 8‑byte length and then the payload.
+5. Reconstruct `DistinctIndex` from newline‑delimited strings.
+
+### Extending DataFusion’s `TableProvider`
+
+We implement a `DistinctIndexTable` that:
+
+* During `try_new`, scans a directory of Parquet files and reads each file’s `DistinctIndex`.
+* In `scan()`, inspects pushed‑down filters on the indexed column (e.g., `category = 'foo'`).
+* Uses `distinct_index.contains(value)` to pick only matching files.
+* Delegates actual row‑group reads to `ParquetSource` for the selected files.
+
+### Usage Overview
+
+```rust
+// Write sample files with embedded indexes
+tmp_dir.iter().for_each(|(name, vals)| {
+    write_file_with_index(&dir.join(name), vals).unwrap();
+});
+
+// Register provider and query
+let provider = Arc::new(DistinctIndexTable::try_new(dir, schema.clone())?);
+ctx.register_table("t", provider)?;
+
+// Only files containing 'foo' will be scanned
+let df = ctx.sql("SELECT * FROM t WHERE category = 'foo'").await?;
+df.show().await?;
+```
+
+---
+
+## Verifying Compatibility with DuckDB
+
+Even with the extra bytes and unknown footer key, standard readers ignore our index. For example:
+
+```sql
+SELECT * FROM read_parquet('/tmp/parquet_index_data/*');
+┌──────────┐
+│ category │
+│ varchar  │
+├──────────┤
+│ foo      │
+│ bar      │
+│ foo      │
+│ baz      │
+│ qux      │
+│ foo      │
+│ quux     │
+│ quux     │
+└──────────┘
+```
+
+DuckDB’s `read_parquet()` sees only the data pages and footer it understands—our embedded index is simply ignored, demonstrating seamless compatibility.
+
+---
+
+## Conclusion
+
+By embedding a custom distinct‑value index directly into Parquet files, we achieve:
+
+* **File‑level pruning** with zero external index files.
+* **Full format compatibility** with standard tools.
+* **Minimal operational overhead**—no special catalog or sidecar management.
+
+This technique illustrates how Parquet’s extensibility can be harnessed for powerful, lightweight indexing, all within the existing format. Give it a try in your next DataFusion project!
```

Comment:
This might be a good place to also mention that parquet itself has many more optimization opportunitues (sort, row group size, etc) that is currently in the introduction

Reply:
Thank you @alamb for this good suggestion!